Cover: Always allow transform from Group block, add handling for Row and Stack variations

andrewserong opened the pull request:
<!-- Thanks for contributing to Gutenberg! Please follow the Gutenberg Contributing Guidelines:
https://github.com/WordPress/gutenberg/blob/trunk/CONTRIBUTING.md -->

Fixes #40209 

## What?
<!-- In a few words, what is the PR actually doing? -->

As raised in #40209, remove the condition that the transform from a Group block to a Cover block option is conditional on a background color being present. Also, add handling to preserve the Stack and Row variations by wrapping those blocks in a Cover block, instead of converting directly to the cover block.

## Why?
<!-- Why is this PR necessary? What problem is it solving? Reference any existing previous issue(s) or PR(s), but please add a short summary here, too -->

Without this option, users wishing to move blocks from a Group to a Cover block need to copy / paste / drag blocks around, so this should make it a little easier to transition a set of blocks over to the Cover block.

## How?
<!-- How is your PR addressing the issue at hand? What are the implementation details? -->

* Remove the condition that a Group block must have a background color in order to be eligible to transform to the Cover block
* Set a default opacity (`dimRatio`) of `50%` if there is no background color — this is needed because the default background is black, and if we otherwise set opacity to `0` then it would not be obvious how to get the color to appear once a user does add a color. Let me know if you can think of a better default other than this medium gray (it's consistent with the settings for if someone removes an image from a cover block).

## Testing Instructions
<!-- Please include step by step instructions on how to test this PR. -->
<!-- 1. Open a Post or Page. -->
<!-- 2. Insert a Heading Block. -->
<!-- 3. etc. -->

1. Add a few blocks to a post or page (e.g. a heading, a paragraph, a button)
2. Select the blocks and group them
3. With the group block selected, select to transform to a Cover block
4. Repeat the above steps with the Row and Stack variations
5. Make sure it still works if the Group block has a background color or gradient set

## Screenshots or screencast <!-- if applicable -->

| Transform from Group block | Transform from Row variation |
| --- | --- |
| ![transform-from-group](https://user-images.githubusercontent.com/14988353/162677847-d9d0e3d0-239d-4941-85cb-f3d75fcc20db.gif) | ![transform-from-row](https://user-images.githubusercontent.com/14988353/162678589-41fb3d96-4645-4d4a-a5a8-a98d33575d50.gif) |

Note: I've used the Bug label as this _feels_ like a UX issue, but could arguably be an enhancement.

andrewserong committed:
Cover: Always allow transform from Group block, add handling for Row and Stack variations

aaronrobertshaw reviewed (commented):
Thanks for putting this together @andrewserong. These transforms look pretty handy for end users. 👍 

My initial testing went as follows:


https://user-images.githubusercontent.com/60436221/162721842-e20c813a-e6b7-47d4-b015-9029ea85a338.mp4



For a standard Group block, this works great! I had slightly more mixed results when testing with the Row or Stack variations. 

If they had a flat background color the resulting Cover block looked as I'd have expected (with the very minor exception of some squashing of Row inner blocks). 

When they have a gradient background, we get a full gradient overlay for the Cover block but the wrapped Group block variation doesn't fill the Cover block. This looks a bit odd and I'm not sure how easy or intuitive it will be for users to correct.

Once we've detected a flex layout variant, could we tweak more in the way of layout attributes to improve this?


https://user-images.githubusercontent.com/60436221/162722586-35cf84b9-aba3-44c6-ada5-37f4eb41843a.mp4

Also, once you transform a Group to a Cover block, going back to the transforms menu again, it appears like you can convert "back" to the original Group block. Performing this transform repeatedly results in nested Cover blocks. 

Is it feasible for us to setup transforms back in the other direction?

I don't believe any of the above is critical if you'd prefer to explore refining these aspects separately.

ntsekouras commented on packages/block-library/src/cover/transforms.js:
```diff
@@ -64,36 +64,54 @@ const transforms = {
+				};
+
+				// For variations that use a flex layout (e.g. Row and Stack),
+				// wrap the block in a Cover instead of converting directly to the cover block.
```
I think some rewording like `preserve the Group block` or something might be a bit more clear.

Also I'm wondering if we should be consistent here and always add the `Group` block nested in the new `Cover` block. That way we can keep the other layout settings.

Finally I think it makes sense to remove existing background related attributes, if they exist, from the Group, as they are also added to the `Cover`. For example if we have a Group with background `red`, it will result in both Group and Cover to have the red background. Makes sense?

andrewserong committed:
Move background color or gradient settings to parent, preserve Group block for all variations

andrewserong commented:
Thanks for the feedback @aaronrobertshaw and @ntsekouras!

I think by doing the following I've addressed most of the issues, but let me know if I missed anything 🙂

* Take background color or gradient settings from the Group block, and apply to the parent Cover block in the transformation, remove from the Group block. (Seems to fix the padding issue on the Row block, too)
* In all cases nest the Group block within the Cover block so that layout is preserved.

| Group w/ gradient to Cover | Row w/ gradient to Cover |
| --- | --- |
| ![group-gradient-to-cover](https://user-images.githubusercontent.com/14988353/162887998-d71ef3c0-2648-48c5-842d-2ebee49b0b08.gif) | ![row-gradient-to-cover](https://user-images.githubusercontent.com/14988353/162888108-4ef15191-b082-4e2f-9556-54ed4ecd254d.gif) |

> Is it feasible for us to setup transforms back in the other direction?

I started having a look at this, too, but it also involves making a few decisions about when to treat the inner group as the block to transform to, and when to preserve it. Since I'll be going AFK after tomorrow, I might look into that in a separate PR so as not to hold up getting this fix in. (That way I can leave it as a WIP if it isn't ready before I go AFK).

Thanks again!

aaronrobertshaw reviewed (approved):
Nice work and quick turnaround on these fixes 🚀 

> Take background color or gradient settings from the Group block, and apply to the parent Cover block in the transformation, remove from the Group block. 

This fixes the issue in the resulting Cover block for Group's with a gradient background. 👍 

> (Seems to fix the padding issue on the Row block, too)

Removing the background from the wrapped Group block does remove the [default padding](https://github.com/WordPress/gutenberg/blob/b12e82a53d534eac76550e5d03d513f8458b4cf2/packages/block-library/src/group/theme.scss) that is applied to Group blocks with the `has-background` class. 

I think that might be all we can do. Some blocks, e.g. Buttons, within a Row block will eat up excessive space and possibly cause some wrapping.

> In all cases nest the Group block within the Cover block so that layout is preserved.

✅ When converting a Group, Row, or Stack, they are all preserved within the newly created Cover block.

> I might look into that in a separate PR

Sounds like a plan. With that in mind, I'd say this is good to go 🚢 


https://user-images.githubusercontent.com/60436221/162891892-5ab794dc-b527-468c-b0cd-e986197348c5.mp4

andrewserong commented:
Thanks for reviewing, Aaron! 🙇
I'll merge this in now, and see how far I get with a PR for going the other direction (Cover -> Group) before I go AFK.

andrewserong commented:
I've opened a follow-up PR to explore going from Cover -> Group in https://github.com/WordPress/gutenberg/pull/40293 — I'm about to head off on leave, so I likely won't see any replies on it, but anyone's welcome to take over that PR if they'd like to. Otherwise, I'll be happy to keep chipping away in a couple of weeks!

adamziel commented:
This was cherry-picked to the `release/13.0` branch to be included in the Gutenberg 13.0 stable release.
